pmlc/dialect/comp/ir/ops.td

```diff
@@ -0,0 +1,267 @@
+// include "mlir/Dialect/LLVMIR/LLVMOpBase.td"
+
+include "interfaces.td"
+
+// TODO Type constraints
+// TODO Verification
+// TODO Builders
+// TODO Descriptions
+
+// =============================================================================
+// Dialect
+// =============================================================================
+def COMP_Dialect : Dialect {
+  let name = "comp";
+  let cppNamespace = "pmlc::dialect::comp";
+  let extraClassDeclaration = [{
+    Type parseType(DialectAsmParser& parser) const override;
+    void printType(Type, DialectAsmPrinter& printer) const override;
+  }];
+}
+
+// =============================================================================
+// Types
+// =============================================================================
+def COMP_IsExecEnvTypePred : CPred<"$_self.isa<::pmlc::dialect::comp::ExecEnvType>()">;
+def COMP_AnyExecEnv : Type<COMP_IsExecEnvTypePred>;
+
+def COMP_IsEventTypePred : CPred<"$_self.isa<::pmlc::dialect::comp::EventType>()">;
+def COMP_AnyEvent : Type<COMP_IsEventTypePred>;
+
+// =============================================================================
+// Operations
+// =============================================================================
+class COMP_Op<string mnemonic, list<OpTrait> traits = []> :
+    Op<COMP_Dialect, mnemonic, traits>;
+
+def COMP_CreateExecEnv : COMP_Op<"create_execenv",
+                                 [MemoryEffects<[MemAlloc<DefaultResource>]>]>
+                       , Results<(outs COMP_AnyExecEnv:$execEnv)> {
+  let summary = "Creates execution environment.";
+
+  let description = [{
+    Execution environment provides functionality for scheduling operations and
+    synchronizing between them.
+  }];
+
+  let assemblyFormat = [{
+    attr-dict `:` type($execEnv)
+  }];
+}
+
+def COMP_DestroyExecEnv : COMP_Op<"destroy_execenv",
+                                  [MemoryEffects<[MemFree<DefaultResource>]>]> {
+  let summary = "Destroys execution environment.";
+
+  let description = [{
+    Any usage of environment or connected with it memory and events is invalid
+    after this operation.
+    Does not guarantee that submitted tasks finished executing.
+  }];
+
+  let arguments = (ins COMP_AnyExecEnv:$execEnv);
+
+  let assemblyFormat = [{
+     $execEnv attr-dict `:` type($execEnv)
+  }];
+}
+
+def COMP_ScheduleEnd : COMP_Op<"schedule_end", [Terminator]> {
+  let summary = "End of scheduling block.";
+}
+
+def COMP_ScheduleSingleBlock : SingleBlockImplicitTerminator<"ScheduleEnd">;
+
+def COMP_ScheduleFunc : COMP_Op<"schedule_func",
+                                [COMP_ScheduleSingleBlock,
+                                 DeclareOpInterfaceMethods<COMP_ExecEnvOpInterface>,
+                                 DeclareOpInterfaceMethods<COMP_ScheduleOpInterface>]>
+                  , Arguments<(ins COMP_AnyExecEnv:$execEnv,
+                                   Optional<COMP_AnyEvent>:$depEvent)>
```

Comment:
Perhaps we want to have a `Variadic` here to support multiple dependent events.

Reply:
Good point, previously I used solution with `group_events` and "empty" event in case I would need variadic for something else. I created more polished solution as #1282